### PR TITLE
New AST types

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Introduce new AST type; use it to replace current uses of AstExtended; for now edits and synchronization are implemented on the old mechanisms (text edits / RelativeRange id map).

### Important Notes

- Edit-related code is commented out until the next PR because it is incompatible with the transitional IdMap-based synchronization.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] ~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~
- [ ] ~All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.~
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] ~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

### CODE REVIEW

File: app/gui2/src/util/ast/abstract.ts
Comment (Frizi) [inline app/gui2/src/util/ast/abstract.ts]: I'd drop a "new" from the docs here, since it will not be new anymore once we have it merged and used.

File: app/gui2/src/util/ast/abstract.ts
Comment (Frizi) [inline app/gui2/src/util/ast/abstract.ts]: I'm not sure if this will necessarily be the only use. The concrete AST might still be more suited for things like alias analysis, unless you think that it can be easily implemented on top of this abstract AST as well.

File: app/gui2/src/util/ast/abstract.ts
Comment (Frizi) [inline app/gui2/src/util/ast/abstract.ts]: Let's move that test into its own test file. We will surely have more than that added later as well.

**Comment by kazcw** [comment]
**`abstract.ts` is ready for review**--I don't expect any further changes, except to the editing functionality which is currently commented. WIP changes to the application logic are included to demonstrate usage of the new APIs.

The rest of the API changes, except for using the new AST for editing, will be ready soon (hopefully today). Editing will be a follow-up PR.

File: app/gui2/src/stores/graph/index.ts
Comment (kazcw) [inline app/gui2/src/stores/graph/index.ts:386]: Is there currently any way to reach this?

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: The difference between `Token` and `Tok` is not easily recognizable. I'd prefer renaming (if possible in JS) the `Token` from generated AST as `RawToken` and rename `Tok` to `Token` (it will be then obvious what is the "main thing")

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: Why some fields and methods starts from underscore? Why `_code` is not just a readonly property?

Browsing the code further, I have no clue about the "start-with-underscore" policy. Those clearly aren't unused things.

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: What about making it parameterized? Like `NodeChild<T = AstId | Tok>` and replace `AstWithWhitespace`/`TokWithWhitespace` with `NodeChild<AstId>`/`NodeChild<Tok>`

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: We may get once, and check if `=== undefined`.

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: Initially, I had no idea why this code works. As I see, `abstract` does not set any of `edited` or `commited`. Realizing where actually globals are affected requires some investigation.

In general, this "global" approach a bit eccentric. Why do we need this? Is it some sort of optimization? What if I would like to parse two different codes which may share ID? What if someone modifies `abstract` code so it won't call any constructor? Not to mention, we resign this way from JS GC and must clear the memory by hand.

Even if we keep this, the code should be heavy of comments: here we need an explanation why `!` is safe.; `abstract` has to have all guarantees documented and ensured by tests.

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: Let's make a better name here. Like `print_subnode` or something along.

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: When do we plan to address this TODO?

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: Why both `NegationOprApp` and `PropertyAccess` does not make any checks of opr tokens? I guess negation should only work with `-` and property access with `.`.

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: What does it represent? When I would use it?

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts:607]: Cannot we give it some shape? Like fields for name, `=` token and the argument itself?

File: app/gui2/src/util/ast/abstract.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/abstract.ts]: What is responsibility difference between `abstract` and `abstract_`?

**Review by farmaazon — COMMENTED** [review]
First set of comments. I've reviewed abstract.ts mostly.

**Comment by kazcw** [comment]
I will be stacking this on #8339, so there's no need for anyone to review the current widgets changes

File: app/gui2/src/stores/graph/graphDatabase.ts
Comment (farmaazon) [inline app/gui2/src/stores/graph/graphDatabase.ts]: TODO left.

File: app/gui2/src/stores/graph/index.ts
Comment (farmaazon) [reply app/gui2/src/stores/graph/index.ts:386]: > Is there currently any way to reach this?

Currently, not. It should be possible after #8068

File: app/gui2/src/stores/graph/index.ts
Comment (farmaazon) [inline app/gui2/src/stores/graph/index.ts]: TODO left

I think the code was bad anyway. The expression ID is a _call_ expression - we should get method pointer from expression updates and this way find the definition.

You may leave my note as comment here for #8068 implementor.

File: app/gui2/src/util/ast/index.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/index.ts]: This is called "post order", but only the root node is yielded in post order.

BTW as calling `next(false)` does not make sense (we've already visited the children) we could make it just a Generator instead of visitor.

File: app/gui2/src/util/ast/token.ts
Comment (farmaazon) [inline app/gui2/src/util/ast/token.ts]: Do we need to do it by classes? We could just have a structure (or a single class) with token type.

**Review by farmaazon — CHANGES_REQUESTED** [review]
Rest of the review.

**Review by farmaazon — APPROVED** [review]
I haven't checked if my comments are resolved, but we agreed to merge it to unblock work. I will re-review the AST-related code once it will be more or less done (with editing).

**Review by Frizi — APPROVED** [review]

**Pull request merged**
